Tripal 4 -  FASTA loader

risharde opened the pull request:
Working version of FASTA loader with included tests / fixtures - up for review. Seems to pass most of the automated tests.

Some tested in regard to the form for specifying additional fields would be nice since I only did a basic test with the Citrus sinensis files from the Tripal 3 documentation.

@spficklin @laceysanderson

risharde committed:
Fasta lImporter T3 code that needs upgrading

risharde committed:
Adjust log errors, upgrade selects and inserts

risharde committed:
Replace form_set_error with Drupal messenger addError function

risharde committed:
Latest changes

risharde committed:
Latest changes

risharde committed:
Latest updates

risharde committed:
FASTA tests beginning

risharde committed:
Latest changes

risharde committed:
Additional tests and fixture for FASTA loader

dsenalik commented:
I wrote documentation on how to create a docker from an unmerged branch here https://github.com/tripal/tripal_doc/pull/34

I used this docker to test. My file loaded okay, except when it was too big (over quota) and I didn't notice. The loader lets you continue even if no fasta file is selected, then you get errors. We need a validation that a file has been selected.
![fasta-import-no-file](https://user-images.githubusercontent.com/8419404/233848016-b162ea8d-46b1-4b42-883a-2967de9fffcc.png)

laceysanderson committed:
Merge branch '4.x' into tv4g4-issue1460-FASTALoader

laceysanderson committed:
Merge branch '4.x' into tv4g4-issue1460-FASTALoader

laceysanderson commented on tripal_chado/src/Plugin/TripalImporter/FASTAImporter.php:
```diff
@@ -0,0 +1,1243 @@
+      // TRIPAL 3 OLD CODE
+      // '#autocomplete_path' => "admin/tripal/storage/chado/auto_name/cvterm/$cv_id",
+      // TRIPAL 4 CODE
+      '#autocomplete_route_name' => 'tripal.cvterm_autocomplete',
```
You'll want to use the new autocomplete contributed by @reynoldtan in PR #1486 instead of this one. This one is restricted to Tripal Terms, whereas the chado version in the PR would select from any cvterm in chado.

Note: I'm not sure the "restrict to cv_id" functionality has been completed yet 🤔 so you'll want to check that and if not, it will need to be implemented (preferably in a separate PR) before this one can be merged.

laceysanderson reviewed (changes requested):
Code review of FASTA importer and tests look good except for the change needed to the autocomplete :-) So once that's fixed and you've added the check for input file in the form validate as mentioned by @dsenalik, then this should be mergable.

laceysanderson committed:
Merge branch '4.x' into tv4g4-issue1460-FASTALoader

risharde committed:
Added FASTA file upload validation

risharde committed:
Use tripal_chado cvterm autocomplete

risharde committed:
Remove debug code

risharde commented:
@laceysanderson Apologies for the delay on this, I finally got around to fixing the 2 things needed for the FASTA loader.
1. Added file validation to detect when no existing file or no new file has been uploaded.
2. Replaced my autocomplete with the recommended tripal_chado cvterm autocomplete code.

Hopefully we are ready for this one to be merged?

dsenalik committed:
Merge branch '4.x' into tv4g4-issue1460-FASTALoader

dsenalik committed:
changes for php 8.2 docker 10.1

dsenalik reviewed (changes requested):
I built a new bleeding-edge docker to test, PHP 8.2, Drupal 10.1

 1. There is a bug in the cvterm autocomplete as I mentioned for the project autocomplete controller here https://github.com/tripal/tripal/pull/1554#issuecomment-1616003506
If you enter for example "zzz" the spinning circle never stops
![zzz](https://github.com/tripal/tripal/assets/8419404/b82da04b-f181-47db-8a2e-0f8d0602dce4)
and the fix is super simple, so I am just going to 
 - [x] add this one change to this branch https://github.com/tripal/tripal/pull/1488/commits/60447c8d1aeb30129ed5559dbaa5c69b2c91b5be

 2. Okay, test loading a FASTA file. I wanted to select DNA as the type, but only one option showed up in the autocomplete
![noDNA](https://github.com/tripal/tripal/assets/8419404/7259c5d1-52be-4e95-a8f7-864efa31558e)
This is a bug in the autocomplete, it is missing a `LOWER()` in the SQL.
 - [x] I might as well add that fix also https://github.com/tripal/tripal/pull/1488/commits/60447c8d1aeb30129ed5559dbaa5c69b2c91b5be

 3. With that fixed, I tried loading but did not specify a file yet. I got
`Warning: Undefined array key "file_upload_existing" in Drupal\tripal_chado\Plugin\TripalImporter\FASTAImporter->formValidate() (line 321 of /var/www/drupal9/web/modules/contrib/tripal/tripal_chado/src/Plugin/TripalImporter/FASTAImporter.php) `
 - [x] so let's include that fix too! https://github.com/tripal/tripal/pull/1488/commits/60447c8d1aeb30129ed5559dbaa5c69b2c91b5be

 4. Okay, finally upload a file and test. I thought it wasn't working because I saw this (and I only read as far as the error message and thought it had failed)
![nofasta](https://github.com/tripal/tripal/assets/8419404/a99113fa-fc45-43a9-ac15-5a9d3c2288f9)
 - [x] ~so this **needs to be fixed**~  Fixed with https://github.com/tripal/tripal/commit/f87be1f044a9d7b203682469254e4390f2559a73

 5. But the job had been submitted successfully. However, when running the job I get
`[error]  Error: Call to undefined method Drupal\Core\Extension\ModuleHandler::getImplementations() in tripal_run_importer() (line 122 of /var/www/drupal9/web/modules/contrib/tripal/tripal/src/api/tripal.importer.api.php) #0 [internal function]: tripal_run_importer('10', Object(Drupal\tripal\Services\TripalJob))`
- [ ] This **will be fixed** in pull #1558

dsenalik commented:
Today's update, point 4 above, I am not getting a file specified in the form state on submit, note that `file_upload` is an empty string.
```
form_state_values =>

array:24 [▼
  "importer_plugin_id" => "chado_fasta_loader"
  "file_upload" => ""
  "file_upload_table_key" => "0-tripal_importer-tripal"
  "file_upload_table" => ""
  "file_upload_submit" => "Upload File"
  "analysis_id" => "1"
  "schema_name" => "chado"
  "organism_id" => "1"
  "seqtype" => "DNA (SO:0000352)"
  "method" => "2"
  "match_type" => "1"
  "re_help" => Drupal\Core\StringTranslation\TranslatableMarkup {#329 ▶}
  "re_name" => ""
  "re_uname" => ""
  "re_accession" => ""
  "db_id" => ""
  "rel_type" => ""
  "re_subject" => ""
  "parent_type" => ""
  "button" => Drupal\Core\StringTranslation\TranslatableMarkup {[#299 ▶](http://localhost:8081/admin/tripal/loaders/chado_fasta_loader#sf-dump-1962370721-ref2299)}
  "form_build_id" => "form-3OjvSQucfQWEBBW2Ivoad71ZZL71OBOHH07-aAVE6Ns"
  "form_token" => "-vrTOl_gTLQ1zGlGTNGQRRSqbLp-3rMXrAEj3haeds0"
  "form_id" => "tripal_admin_form_tripalimporter"
  "op" => Drupal\Core\StringTranslation\TranslatableMarkup {[#299 ▶](http://localhost:8081/admin/tripal/loaders/chado_fasta_loader#sf-dump-1962370721-ref2299)}
]
```

But maybe that's because I need to hit the "Upload File" button. But when I do that I get
![500error](https://github.com/tripal/tripal/assets/8419404/460b625b-616f-4e95-bfd0-ca8ac95671bc)

and in the log there are three of
`Error: Call to undefined method Drupal\Component\Utility\Bytes::toInt() in tripal_file_upload() (line 193 of /var/www/drupal9/web/modules/contrib/tripal/tripal/src/api/tripal.upload.api.php).`

**Bytes::toInt was deprecated in Drupal 9 and removed in Drupal 10.**
https://www.drupal.org/project/drupal/issues/3142934

But fix that and then:
```Error: Undefined constant "FILE_STATUS_PERMANENT" in tripal_file_upload_merge() (line 505 of /var/www/drupal9/web/modules/contrib/tripal/tripal/src/api/tripal.upload.api.php).```

**FILE_STATUS_PERMANENT is deprecated**
https://www.drupal.org/node/3022147

risharde commented:
Strange, I tested this before the latest 4.x changes when I patched for the
validation and this seemed to work. I'll have to look into this further but
it's odd that it all of a sudden doesn't show file_upload => <file_id>
which it should show if it's a new FASTA file uploaded. The validation
checks this key to see if it was a new file uploaded or not. It also checks
to see if alternatively a file existing is set and should use that - this
was tested on the weekend after a cache clear but maybe some update broke
it.

I'll report back when I check on this - hopefully some time this week if
you don't patch it before that, thanks for testing!

On Mon, Jul 3, 2023 at 12:37 PM Douglas Senalik ***@***.***>
wrote:

> Today's update, point 4 above, I am not getting a file specified in the
> form state on submit, note that file_upload is an empty string.
>
> form_state_values =>
>
> array:24 [▼
>   "importer_plugin_id" => "chado_fasta_loader"
>   "file_upload" => ""
>   "file_upload_table_key" => "0-tripal_importer-tripal"
>   "file_upload_table" => ""
>   "file_upload_submit" => "Upload File"
>   "analysis_id" => "1"
>   "schema_name" => "chado"
>   "organism_id" => "1"
>   "seqtype" => "DNA (SO:0000352)"
>   "method" => "2"
>   "match_type" => "1"
>   "re_help" => Drupal\Core\StringTranslation\TranslatableMarkup {#329 ▶}
>   "re_name" => ""
>   "re_uname" => ""
>   "re_accession" => ""
>   "db_id" => ""
>   "rel_type" => ""
>   "re_subject" => ""
>   "parent_type" => ""
>   "button" => Drupal\Core\StringTranslation\TranslatableMarkup {[#299 ▶](http://localhost:8081/admin/tripal/loaders/chado_fasta_loader#sf-dump-1962370721-ref2299)}
>   "form_build_id" => "form-3OjvSQucfQWEBBW2Ivoad71ZZL71OBOHH07-aAVE6Ns"
>   "form_token" => "-vrTOl_gTLQ1zGlGTNGQRRSqbLp-3rMXrAEj3haeds0"
>   "form_id" => "tripal_admin_form_tripalimporter"
>   "op" => Drupal\Core\StringTranslation\TranslatableMarkup {[#299 ▶](http://localhost:8081/admin/tripal/loaders/chado_fasta_loader#sf-dump-1962370721-ref2299)}
> ]
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tripal/tripal/pull/1488#issuecomment-1618858231>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AICMFI3FVZNB36JFYUFA2PDXOLYKZANCNFSM6AAAAAAXHTL6YU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

dsenalik committed:
Fix two deprecations

dsenalik commented:
Okay I added two more fixes with commit https://github.com/tripal/tripal/pull/1488/commits/f87be1f044a9d7b203682469254e4390f2559a73, and (provided I temporarily include pull #1558) everything worked and I was able to import using PHP 8.2, Drupal 10.1

risharde commented:
Wonderful, thanks Doug!

On Mon, Jul 3, 2023 at 1:05 PM Douglas Senalik ***@***.***>
wrote:

> Okay I added two more fixes with commit f87be1f
> <https://github.com/tripal/tripal/commit/f87be1f044a9d7b203682469254e4390f2559a73>,
> and (provided I temporarily include pull #1558
> <https://github.com/tripal/tripal/pull/1558>) everything worked and I was
> able to import using PHP 8.2, Drupal 10.1
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tripal/tripal/pull/1488#issuecomment-1618896430>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AICMFI7K52F3RNW7HA4J74LXOL3WHANCNFSM6AAAAAAXHTL6YU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

laceysanderson commented:
Awesome, thanks for the review on this Doug! 
Note: I re-requested your review because you had left it as "changes requested" after comment https://github.com/tripal/tripal/pull/1488#pullrequestreview-1509624678 but then not dismissed/approved based on comment https://github.com/tripal/tripal/pull/1488#issuecomment-1618896430 which looks like all is well now based on your fixes?

I also re-requested a review from myself but will make sure to look at it after merging PR #1558.

dsenalik reviewed (approved):
Now approved

dsenalik committed:
Remove temp file

dsenalik committed:
include commit 923b9b9 from pull 1554

dsenalik committed:
Merge branch '4.x' into tv4g4-issue1460-FASTALoader

laceysanderson committed:
Merge branch '4.x' into tv4g4-issue1460-FASTALoader

laceysanderson reviewed (approved):
Changes look good ✅ 

Although we will want to filter the cvterm autocomplete to the sequence ontology (now added as a separate issue #1571), I did confirm that if you choose a term from another vocabulary you do get a validation error at the form level to help you fix it. In the interests of getting this merged, I'm leaving the actual restriction of the autocomplete to issue #1571.

I also test loaded files and it all still works as expected 👍 

Thanks @risharde, this is ready to merge 🎉
